File: python/wakeword.py
## Set Initial Variables ##
import os # Miscellaneous operating system interface
import zmq # Asynchronous messaging framework
import time # Time access and conversions
import sys # System-specific parameters and functions
from matrix_io.proto.malos.v1 import driver_pb2 # MATRIX Protocol Buffer driver library
from matrix_io.proto.malos.v1 import io_pb2 # MATRIX Protocol Buffer io library
from multiprocessing import Process # Allow for multiple processes at once
from zmq.eventloop import ioloop # Asynchronous events through ZMQ
matrix_ip = '127.0.0.1' # Local device ip
wakeword_port = 60001 # Driver Base port
# Handy functions for connecting to the Data Update, & Error port 
from utils import register_data_callback, register_error_callback
# Sphinx Knowledge Base files
LM_PATH = 'INSERT_PATH_TO_YOUR_FILE.lm'# Language Model File
DIC_PATH = 'INSERT_PATH_TO_YOUR_FILE.dic'# Dictation File

## BASE PORT ##
def config_socket():
    # Define zmq socket
    context = zmq.Context()
    # Create a Pusher socket
    socket = context.socket(zmq.PUSH)
    # Connect Pusher to configuration socket
    socket.connect('tcp://{0}:{1}'.format(matrix_ip, wakeword_port))
    
    # Create a new driver config
    config = driver_pb2.DriverConfig()
    # Language Model File
    config.wakeword.lm_path = LM_PATH
    # Dictation File
    config.wakeword.dic_path = DIC_PATH
    # Desired MATRIX microphone
    config.wakeword.channel = 8
    # Enable verbose option
    config.wakeword.enable_verbose = False

    # Send driver configuration through ZMQ socket
    socket.send(config.SerializeToString())
    print ('Listening for wakewords')

## KEEP-ALIVE PORT ##
def ping_socket():
    # Define zmq socket
    context = zmq.Context()
    # Create a Pusher socket
    ping_socket = context.socket(zmq.PUSH)
    # Connect to the socket
    ping_socket.connect('tcp://{0}:{1}'.format(matrix_ip, wakeword_port+1))
    # Send a ping every 2 seconds
    while True:
        ping_socket.send_string('')
        time.sleep(2)

## ERROR PORT ## (!Currently Experiencing False Errors!)
def wakeword_error_callback(error):
    # Log error
    print('{0}'.format(error))

## DATA UPDATE PORT ##
def wakeword_data_callback(data):
    # Extract data
    data = io_pb2.WakeWordParams().FromString(data[0])
    # Log data 
    print('{0}'.format(data))
    # Run actions based on the phrase heard
    # CHANGE TO YOUR PHRASE
    if data.wake_word == 'MATRIX START':
        print ('I HEARD MATRIX START!\n')
    # CHANGE TO YOUR PHRASE
    elif data.wake_word == 'MATRIX STOP':
        print ('I HEARD MATRIX STOP!\n')

## Start Processes ##
if __name__ == '__main__':
    # Initiate asynchronous events
    ioloop.install()
    # Send Base Port configuration 
    config_socket()
    # Start Error Port connection
    # Process(target=register_error_callback, args=(wakeword_error_callback, matrix_ip, wakeword_port)).start()
    # Start Data Update Port connection
    Process(target=register_data_callback, args=(wakeword_data_callback, matrix_ip, wakeword_port)).start()
    # Start Keep-alive Port connection
    Process(target=ping_socket).start()

File: javascript/wakeword.js
// Set Initial Variables \\
var zmq = require('zeromq');// Asynchronous Messaging Framework
var matrix_io = require('matrix-protos').matrix_io;// MATRIX Protocol Buffers
var matrix_ip = '127.0.0.1';// Local Device IP
var matrix_wakeword_base_port = 60001; // Wakeword base port
const LM_PATH = 'INSERT_PATH_TO_YOUR_FILE.lm';// Language Model File
const DIC_PATH = 'INSERT_PATH_TO_YOUR_FILE.dic';// Dictation File

// BASE PORT \\
// Create a Pusher socket
var configSocket = zmq.socket('push');
// Connect Pusher to Base port
configSocket.connect('tcp://' + matrix_ip + ':' + matrix_wakeword_base_port /* config */);
// Create driver configuration
var config = matrix_io.malos.v1.driver.DriverConfig.create(
{ // Create & Set wakeword configurations
  wakeword: matrix_io.malos.v1.io.WakeWordParams.create({
    lmPath: LM_PATH,// Language model file path
    dicPath: DIC_PATH,// Dictation file path
    channel: matrix_io.malos.v1.io.WakeWordParams.MicChannel.channel8,// Desired MATRIX microphone
    enableVerbose: false// Enable verbose option
  })
});
// Send configuration to MATRIX device
configSocket.send(matrix_io.malos.v1.driver.DriverConfig.encode(config).finish());
console.log('Listening for wakewords');

// KEEP-ALIVE PORT \\
// Create a Pusher socket
var pingSocket = zmq.socket('push');
// Connect Pusher to Keep-alive port
pingSocket.connect('tcp://' + matrix_ip + ':' + (matrix_wakeword_base_port + 1));
// Send initial ping
pingSocket.send('');
// Send a ping every 2 seconds
setInterval(function(){
  pingSocket.send('');// Send ping
}, 2000);

// ERROR PORT \\
// Create a Subscriber socket
var errorSocket = zmq.socket('sub');
// Connect Subscriber to Error port
errorSocket.connect('tcp://' + matrix_ip + ':' + (matrix_wakeword_base_port + 2));
// Connect Subscriber to Error port
errorSocket.subscribe('');
// On Message
errorSocket.on('message', function(error_message){
  //console.log('Error received: ' + error_message.toString('utf8'));// Log error
});

// DATA UPDATE PORT \\
// Create a Subscriber socket
var updateSocket = zmq.socket('sub');
// Connect Subscriber to Base port
updateSocket.connect('tcp://' + matrix_ip + ':' + (matrix_wakeword_base_port + 3));
// Subscribe to messages
updateSocket.subscribe('');
// On Message
updateSocket.on('message', function(wakeword_buffer) {
  // Extract message
  var wakeWordData = matrix_io.malos.v1.io.WakeWordParams.decode(wakeword_buffer);
  // Log message
  console.log(wakeWordData);
  // Run actions based on the phrase heard
  switch(wakeWordData.wakeWord) {
    // CHANGE TO YOUR PHRASE
    case "MATRIX START":
      console.log('I HEARD MATRIX START!');
      break;
    // CHANGE TO YOUR PHRASE
    case "MATRIX STOP":
      console.log('I HEARD MATRIX STOP!');
      break;
  }
});
